### DISCUSSION TITLE
Adding a Junction

### DISCUSSION THREAD

**Discussion by Diego-planckian1** [iqm-finland/KQCircuits#80] [Q&A]
Hi,

I'm having some troubles in the insertion of my own junction.

I created a junction geometry and place the .py in the junctions folder. I can see it in the GUI of klayout (Planck SQUID):

![Screenshot from 2024-01-08 15-13-00](https://github.com/iqm-finland/KQCircuits/assets/150356602/61dff48e-2415-4e27-b37e-14a3b3737ca6)

By the way once i try to add it inside an other script (for generating a qubit) it print me the default Manhattan junction. How I can I add my own junction?

This is the dummy code I implemented:
![Screenshot from 2024-01-08 15-11-49](https://github.com/iqm-finland/KQCircuits/assets/150356602/2af0f683-dd32-4438-9c4f-d7ec12533f97)

Where PlanckSQUID is the class of my junction.

**Comment by Qvadi** [comment]
I think the problem here is a bit of naming culture in KQCircuits. Normally the import structure is as follows:
For example, my_element.py (a file under the kqcircuits\elements folder) is something that the user creates, then your import structure you need to follow is this:
`from kqcircuits.elements.my_element import MyElement` 

Note: the character after the 'underscore' is capitalized. 
In your case, the faulty line of code is the planck_junction import. The correct way to import this file is,
`from kqcircuits.junctions.planck_junction import PlanckJunction`

If this is not the right file, then adjust it accordingly.

**Reply by Diego-planckian1** [reply]
To avoid misunderstandings I changed the names:

![Screenshot from 2024-01-09 09-52-39](https://github.com/iqm-finland/KQCircuits/assets/150356602/04610600-8592-4007-9f8e-48a5c6defa51)

![Screenshot from 2024-01-09 09-53-04](https://github.com/iqm-finland/KQCircuits/assets/150356602/f3f51705-7510-44b8-ab50-8a2d35a7e1a4)

Still print the manhattan squid...

**Reply by Diego-planckian1** [reply]
Also if I use one of the default junction in my script (e.g. manhattan single junction), in the output i still have the default manhattan junction!
![Screenshot from 2024-01-09 15-15-44](https://github.com/iqm-finland/KQCircuits/assets/150356602/1782cca1-2321-474d-abc1-3fad80ba41c8)

**Reply by Qvadi** [reply]
There is still a typo in your code, should be `PlanckTransmonCap`. Not sure what else could be wrong.

**Reply by Diego-planckian1** [reply]
PlanckTransmonCap is an element, the capacitance of the transmon, and it is inserted correctly. The problem regards the insertion of a junction.:

I tried to insert my own junction --> give me back the default manhattan  junction.

I substituted PlanckJunction with ManhattanSingleJunction ---> again the default Manhattan SQUID.

What could be mistaken? I don't know i the junction adding follows the same rule of the element adding.

**Reply by nikosavola** [reply]
You can try checking the example in https://github.com/iqm-finland/KQCircuits/blob/main/klayout_package/python/kqcircuits/qubits/swissmon.py. Seems the junction is a special case implemented in the `Qubit` class. You should use `self.produce_squid` for placing the junction. This way the PCell parameter for junction type is obeyed.

**Comment by qpavsmi** [comment]
Maybe this can help (sorry for the delay)

Go to `klayout_package/python/kqcircuits/junctions/__init__.py`. There is a dict called `junction_type_choices` with a list of junctions that can be placed in the qubit. These are given as an option when you want to edit the `junction_type` of a qubit of your choice. Try adding your squid as an entry, it should be a string as shown in GUI, either "Planck SQUID" or "Planck Junction".

Then check out Niko's comment, you should call `produce_squid` with transformation as argument that tells at which coordinate and orientation your squid is placed at. Now try drag-and-dropping your qubit to the layout, and change the parameter `junction_type` from the parameters. Might have to click the option two times because this part of gui is a bit buggy :/ Let me know if you'll see your squid on a qubit with this
